frontend/lib/resolution/can-pass.cpp

```diff
@@ -969,11 +969,13 @@ CanPassResult CanPassResult::canPass(Context* context,
     // so checking here just rules out predictable situations.
 
     if (formalQT.kind() != QualifiedType::TYPE &&
+        actualQT.kind() != QualifiedType::INIT_RECEIVER &&
         isTypeGeneric(context, actualQT))
       return fail(FAIL_GENERIC_TO_NONTYPE); // generic types can only be passed to type actuals
 
     auto got = canInstantiate(context, actualQT, formalQT);
-    if (!got.passes() && formalQT.kind() == QualifiedType::TYPE) {
+    if (!got.passes() && (formalQT.kind() == QualifiedType::TYPE ||
```

Comment:
Actually, I think this is a bug in `canPass`. I suspect we need to try conversions for generic formals even if generic actuals aren't allowed. E.g., I suspect:

```
proc f(x: Parent(?)) {}
f(new Child(int));
```

Should work. I can't test it because we don't have parent / child constructors. I'll try writing another case that catches this.

Reply:
I'm going to re-write using the intermediate variable and open an issue about this.

Reply:
Ok, I've looked deeper into this because I tried opening and issue and realized I wasn't sure what I was talking about (is this the Feynman technique in action?). It turns out, `canInstantiate` already checks for borrowing and nilability, but fails if the type can be passed, but doesn't produce substitutions; this makes sense because that would imply that we just passed a generic actual to a generic formal, which is a no-no in most cases. That's why there's logic there to check if the kind is `TYPE` (which does allow generic actuals), and fall through to the checks.

I've adjusted this logic in a way that makes more sense to me: instead of falling through, I detect a "fail due to no substitutions", and explicitly override it if we `acceptsGenericActuals` is true (that's the condition your original comment notes is written twice). This has the slight effect of saving work for `TYPE` formals, and in my opinion resolves the inconsistency.

Reply:
(see b2ba96e20d5201fd99504690e2bb27f548eb5800).

Reply:
(see b2ba96e20d5201fd99504690e2bb27f548eb5800)